content/posts/2025/building-a-minimal-viable-armv7-emulator.md

```diff
@@ -1,49 +1,216 @@
 ---
-title: "Building a Minimal Viable Armv7 Emulator"
-date: 2025-11-06
-summary: "Emulating ARMv7 is suprisingly easy, even from scratch AND in Rust"
-draft: true
+title: "Building a Minimal Viable Armv7 Emulator from Scratch"
+date: 2025-11-19
+summary: "Emulating armv7 is surprisingly easy, even from scratch AND in Rust"
 tags:
   - arm
   - rust
 ---
 
+{{<callout type="Tip or TLDR - I built a tiny, zero dependency armv7 userspace emulator in Rust">}}
+I wrote a minimal viable armv7 emulator in 1.3k lines of Rust without any
+dependencies. It parses and validates a 32-bit arm binary, maps its segments,
+decodes a subset of arm instructions, translates guest and host memory
+interactions and forwards arm Linux syscalls into x86-64 System V syscalls.
+
+It can run a armv7 hello world binary and does so in 1.9ms (0.015ms for raw
+emulation without setup), while qemu takes 12.3ms (stinkarm is thus ~100-1000x
+slower than native armv7 execution).
+{{</callout>}}
+
+After reading about the process the Linux kernel performs to execute binaries,
+I thought: I want to write an armv7 emulator - [`stinkarm`](https://github.com/xnacly/stinkarm). Mostly to understand the ELF
+format, the encoding of arm 32bit instructions, the execution of arm assembly
+and how it all fits together (this will help me with the JIT for my programming
+language I am currently designing). To fully understand everything: no
+dependencies. And of course Rust, since I already have enough C projects going
```

Comment:
```suggestion
dependencies [^1]. And of course Rust, since I already have enough C projects going
```
Footnote about dependencies. Only saw the footnote later and was confused when you directly after that used clap in `bld_exmpl`